### DISCUSSION TITLE
Support for Weight Attribute from SyGuS 2.1

### DISCUSSION THREAD

**Discussion by yanrujhou** [cvc5/cvc5#11108] [Q&A]
Based on the current SyGuS Language Standard Version 2.1, a new command declare-weight has been added to allow users to define weight attributes. This enables users to annotate terms with weights to guide the solver toward certain solution candidates.

However, there is no API library for this feature mentioned in the cvc5 documentation. Additionally, attempts to run cvc5 with this feature indicate that it is not supported.

Is there clarification available regarding whether cvc5 currently supports the feature allowing users to define weights on terms? If not, is there an alternative method to annotate terms with weights so that the solver can bias toward certain solution candidates?

**Comment by ajreynol** [comment]
Sorry for the long delay in reply.  We are planning to add support for this feature in the future. There was a development branch that had this feature implemented but it has yet to be added to main

FYI @abdoo8080

**Comment by bub27** [comment]
Hello, I'd like to follow up on this question. Since a branch seemed to incorporate weight considerations, would it be possible to integrate it into the main branch, or at least make the branch available?

And if not, are there other tools that incorporate weight considerations?

Thank you in advance for your reply.

Bruno Mermet (https://www.linkedin.com/in/bruno-mermet-4a171626b/)

**Comment by abdoo8080** [comment]
Hi Bruno! Yes, our plan is to break #12619 into smaller PRs and merge them into main over the coming weeks. One thing to flag: #12619 only adds minimal weight functionality, so depending on what you're trying to do, it may not be sufficient on its own.

**Reply by bub27** [reply]
Great ! and thanks a lot for your very quick response.
Our need is the generation of a function whose grammar would have a minimum weight. I don't know if this kind of functionality will be included in the upcoming commit... But any other weight consideration might be also useful.
Bruno

**Reply by abdoo8080** [reply]
Thanks for the details. We plan to support minimum-weight synthesis eventually (no firm timeline yet), and #12619 on its own won't give you that: there's no optimality guarantee. It can get you part of the way as a workaround, but with some caveats.

What #12619 adds is support for weight constraints of the form:
```clojure
(declare-weight w)
...
(synth-fun f ...)
...
(constraint p[(_ w f)])
...
(check-synth)
```
where `p[(_ w f)]` is any arbitrary SyGuS constraint over `(_ w f)`. So, you can put upper or lower bounds on the weight and, combined with `check-synth-next`, iterate toward smaller weights.

Two important caveats:
- The enumeration procedure isn't weight-aware: It doesn't enumerate terms in increasing weight order, so a tight bound prunes the search space but doesn't guide it. Tighter bounds may make solutions harder to find rather than easier.
- Infeasible bounds don't fail: If you ask for a term with a weight below what's achievable, the solver won't report "no solution." It'll time out, exhaust resources, or loop forever. As a work-around, you can start from a known-feasible upper bound and tighten monotonically: each success gives you an improvement, and you keep the best-so-far when the next query hangs.

Adding weight-aware enumeration resolves both caveats and is on the future to-do list.
